### ISSUE TITLE
Bezier curves for automation tracks

### ISSUE THREAD

**Issue by codythecoder** [LMMS/lmms#2381]
So one problem I have with the automation tracks as they are is that they lack control for the endpoints. It's impossible to get the curve to start or end going strait upwards, which can sometimes be what you want to do, and the only way to do this at the moment is with a lot of points. I think that bezier curves would be able to solve this problem, and make the automation tracks easier for newcomers and seasoned users alike.
Here's a rough idea of what I'd imagine it would look like:
![temp](https://cloud.githubusercontent.com/assets/13900631/10094006/aa9a8c32-639d-11e5-818b-4f6f91d93c23.png)
![temp2](https://cloud.githubusercontent.com/assets/13900631/10094013/cc5e8e0e-639d-11e5-8510-1be8ad849a24.png)
![temp3](https://cloud.githubusercontent.com/assets/13900631/10094406/dd38d766-63a2-11e5-9dd2-a3305383f73a.png)

I can do the coding, but what do people think? Anything you would add/change to this?

**Comment by musikBear** [comment]
imo brilliant! And if it has no efficiency issues (and why should it..) then it would be a super addition, but for backward reasons, as a additional feature, and not a replacement. Thats important.

**Comment by BaraMGB** [comment]
The Problem with bezier curves is you can't calculate them. You need a curve where you can get for every x an y coordinate. Bezier got's only start and end points and the control points. That's not enough.

**Comment by Spekular** [comment]
It's possible to draw a bezier curve though, so it must be possible to get
coordinates. Blender does it too, and inkscape. I mean, why would there be
such a thing as a bezier curve if it wasn't a specific equstion/whatever
for a curve? Unless you mean that people could make bezier curves that
overlap vertically, but I'd say limits should prevent that from happening.

**Comment by codythecoder** [comment]
Yeah, I've looked around quite a bit and most stuff says you can't get a y from an x value. And being a cubic this is what I'd expect. What I was thinking was to have some sort of x, y table iteration over a limited number of t values and do some linear extrapolation between points. Using a fine enough t this should be smooth enough for any purpose.

(Note: cubics are technically solvable but search "cubic polynomial roots" on google to see why one wouldn't want to do this.)

**Comment by michaelgregorius** [comment]
You can use [De Casteljau's algorithm](https://en.wikipedia.org/wiki/De_Casteljau's_algorithm) to compute finer and finer versions of the control polygon. If you do this an infinite number of steps the resulting polygons will converge towards the actual curve. Of course, we don't have that much time but it should already give some good results if the recursion is stopped after a finite number of steps. :smile:

The result would then be a polygon of linear segments which can be evaluated at arbitrary positions.

**Comment by codythecoder** [comment]
What I'm doing atm is splitting the curve into a given number of linear segments, and then iterating through to find the segment that a given x is in. I have some equations that find the x and y from a given t, so manually calculating a point seems a bit redundant. I'll probably not use recursion as with both algorithms I can just iterate through a given precision of t to find the point, which gives me a fine enough curve at really small numbers of segments (20 segments is almost indistinguishable from 1000 segments).

**Comment by Spekular** [comment]
When rendering perhaps we could use as many segments as the samplerate,
with an option to be less accurate? I'm assuming this isn't feasible to do
"live".

**Comment by codythecoder** [comment]
Running at 1000 segments didn't *seem* to run that slowly (I can only base this off what I see on the screen, as I'm having problems with Ubuntu drivers), so it would definitely be realistic to chuck in a whole bunch of extra segments for rendering, but I'll probably chuck that in after I've figured out all this GUI stuff.

**Comment by Wallacoloo** [comment]
I'll bet you'd get similar accuracy, but better performance from e.g. 20
segments with quadratic interpolation than with 1000 segments using linear
interpolation.

But strictly speaking, we don't *need* mathematical accuracy. As long as it
produces consistent results and the user is able to tweak the curve until
it sounds how they want it, then the tool is doing its job.

**Comment by musikBear** [comment]
> As long as it produces consistent results and the user is able to tweak the curve until it sounds how they want it, then the tool is doing its job.

:+1: agree

**Comment by michaelgregorius** [comment]
I also agree with @Wallacoloo. There is no need to calculate the bezier curve down to sample level accuracy. For most automation targets using an approximation instead of the 100% solution will likely not be noticeable in a full mix anyway.

@codythecoder It think drawing the bezier curves can be handled completely by Qt, e.g. using `QPainterPath::cubicTo` ([link](http://doc.qt.io/qt-4.8/qpainterpath.html#cubicTo)). So the approximations will likely only be needed for the audio side of things.

**Comment by codythecoder** [comment]
Just in case anyone's interested, this is what it looks like atm
![screenshot from 2015-10-15 11 26 46](https://cloud.githubusercontent.com/assets/13900631/10501699/c687c56c-732f-11e5-9c46-642a79f11cf4.png)

**Comment by musikBear** [comment]
@codythecoder Super! :+1: 
How does the curve behave, if you try to use 'mirror' over x a.o y. Are that feature still supported without problems?

**Comment by Spekular** [comment]
The regular points should mirror just fine, but the handles need to flip as well. It could get a bit tricky depending on how they're stored, whether they're relative to the point or absolute, etc. and as such I'd like to waive all responsibility :stuck_out_tongue_closed_eyes: 
Kidding aside it shouldn't be too bad but my dev environment is gone and I'm a bit short on time atm, so it probably won't be me either way. If we're lucky it'll just work :four_leaf_clover:

**Comment by codythecoder** [comment]
From what I've seen of the mirror function, it shouldn't be too hard to get them working. The flip over y should require almost not coding (yay) but the x might be a bit trickier because of how they're stored.

**Comment by codythecoder** [comment]
Ok, I've basically finished, but I have no idea how to use github to actually put my code on my fork so that other people can test it, so if someone could show me how to do this, that'd be great.
It would be great if someone with some design skills could make an icon for the bezier progression because I'm not sure my design fits in with the rest of the icons.
This is what it looks like at the moment:
![progression_bezier](https://cloud.githubusercontent.com/assets/13900631/10755361/653a216c-7cf1-11e5-90be-111d9263bf3d.png)
and this is what the handles look like with everything:
![screenshot from 2015-10-27 21 27 48](https://cloud.githubusercontent.com/assets/13900631/10755403/b0be8074-7cf1-11e5-912d-33e5c4fd1926.png)

**Comment by michaelgregorius** [comment]
@codythecoder Looks great! Where can we find the current version of your code? I have found a commit in the master branch of your LMMS clone. However, when I tried to compile it it failed with the following error message:
```
[ 20%] Linking CXX executable ../lmms
CMakeFiles/lmmsobjs.dir/lmmsobjs_automoc.cpp.o: In function `AutomationEditor::qt_metacall(QMetaObject::Call, int, void**)':
lmmsobjs_automoc.cpp:(.text+0x8a81): undefined reference to `AutomationEditor::controlPointColor() const'
lmmsobjs_automoc.cpp:(.text+0x8ae9): undefined reference to `AutomationEditor::setControlPointColor(QColor const&)'
collect2: Fehler: ld gab 1 als Ende-Status zurück
src/CMakeFiles/lmms.dir/build.make:484: die Regel für Ziel „lmms“ scheiterte
make[2]: *** [lmms] Fehler 1
CMakeFiles/Makefile2:1069: die Regel für Ziel „src/CMakeFiles/lmms.dir/all“ scheiterte
make[1]: *** [src/CMakeFiles/lmms.dir/all] Fehler 2
Makefile:149: die Regel für Ziel „all“ scheiterte
make: *** [all] Fehler 2
```

**Comment by codythecoder** [comment]
Yeah, that's what I get for not understanding command line git and so copying all the files to windows where I would use the github gui to do the commiting and whatnot.
Anyhow, figured out command line ubuntu git stuff, so it should be working now, with the most recent version.
For some reason all the buttons changed look and I'm not sure why. Everything still works though.

**Comment by codythecoder** [comment]
There's still two bugs (that I know of) and a peculear edge case to watch out for.
 * Clicking directly on an automation point will not move it. This involves directly under it as well, but it has to be *exactly* on or under it.
 * If you move or create an automation point, and you move it to an existing automation point, it will not move the control point, which I think might have been my intentional fault it does this. But I don't like it any more, it feels weird with the rest of it imo.
 * Flipping this track horizontally does what it's supposed to but I'm not sure if this is right?
![screenshot from 2015-10-28 15 54 23](https://cloud.githubusercontent.com/assets/13900631/10780035/03f255fe-7d8d-11e5-90a9-3e4ee6b6068d.png)
![screenshot from 2015-10-28 16 00 52](https://cloud.githubusercontent.com/assets/13900631/10780037/15908cd6-7d8d-11e5-8b1b-0ae1f3a706d4.png)
(Notice one automation point has no visible handles)

**Comment by Spekular** [comment]
I'm pretty sure the second automation point has no handles since it's a relocated first point:
![control points](https://cloud.githubusercontent.com/assets/8918426/10783598/3707f73e-7d59-11e5-80da-5951c58bda85.png)
(Green = has handles, Red= doesn't have handles)

I'm more concerned about the first pattern ending just short of three while the second ends at 2:
![flip](https://cloud.githubusercontent.com/assets/8918426/10783605/48252c4e-7d59-11e5-867e-c6e46caa0afe.png)
(Yellow dotted line is extrapolated handle, which could line up with the endpoint of the post-flip curve)

By "control point" I meant handle. I'm assuming the second point post-flip is being placed where the second point's *first handle* is pre-flip. I hope this makes sense.

**Comment by codythecoder** [comment]
Yeah, I was kinda using a different pattern because I realised I closed LMMS without taking a screenshot with it flipped. I probably should've mentioned that.
![temp](https://cloud.githubusercontent.com/assets/13900631/10784179/c398354a-7db0-11e5-8135-0a1321a4f572.png)

So both this point's control points are outside of the screen, so it's correctly flipped, but now you can't move the control points without moving the automation point.
Since it's correctly flipped, but now you can't grab the control points, I don't know what the "correct" behavior should be.

**Comment by musikBear** [comment]
in respect to *mirror* ao *flip*:
Imo that actually only need to be raw values. The flipped/mirrored patten does not need to have control-points (handles) at all.
I argue this out from the actual functionality, over what we could call 'design-press'
The **usecase** the mirror/flip feature were meant to address, was to have an easy way to *make the exact opposite pattern from an existing one*. The 'inverted' copy was *never intended to be edited* at all.

**Comment by tresf** [comment]
>  The 'inverted' copy was never intended to be edited at all.

Shut up.

**Comment by michaelgregorius** [comment]
@codythecoder I have pulled the updated code and now everything compiles. I also did a quick test and everything seems to work quite well so far.

However, you might want to remove all the debugging statements that are printed to the console before a final merge. :wink:
```
Is here
pos_ticks: 501
_time: 528
Should be here
flipped
controlPoint == 2: 1
pos_ticks: 500
_time: 500
[...]
```

**Comment by codythecoder** [comment]
@musikBear that makes sense. And if fit whatever reason the user does need to move that point they can just move the automation point.
@michaelgregorius yeah, there is a few :smile: 
shouldn't be too hard to remove them all though.

**Comment by tresf** [comment]
Pardon my ignorance here, but are we really ok with this statement:

> The 'inverted' copy was never intended to be edited at all.

**Comment by StakeoutPunch** [comment]
>Pardon my ignorance here, but are we really ok with this statement.

I'm not, as I'm still trying to understand it myself. If he means that a flipped pattern isn't supposed to be edited, then definitely no. What kind of interface "feature" is that?

**Comment by musikBear** [comment]
well it was my idea, i just wanted an inverted carbon-copy so the effect could be exactly repeated, but reversed, just as my own app
https://www.youtube.com/watch?v=yPaOva0ntzs
 does. Imo it would be a shame to miss out of bezier curve advantages, because of the issue of flipping bezier-handles.

> that makes sense. And if fit whatever reason the user does need to move that point they can just move the automation point.

Yes that was what i thought ;)


> Shut up.

Please do not apply as mediator in the middle-east theater :hear_no_evil:   :p

**Comment by codythecoder** [comment]
I think it's just about done, so what's the process of getting it from https://github.com/codythecoder/lmms to the main repository?

**Comment by tresf** [comment]
> what's the process of getting it from https://github.com/codythecoder/lmms to the main repository?

A good start is a [Pull Request](https://github.com/LMMS/lmms/compare/master...codythecoder:master).

I'm still a bit concerned about the questions above.  @codythecoder, @musikBear can you explain in greater detail how this won't be a problem?

> The 'inverted' copy was never intended to be edited at all.

**Comment by musikBear** [comment]
@tresf 
> can you explain in greater detail how this won't be a problem?

I can try.. 
the reason for making a mirror copy, would primary be done as a solution to having an increase automation going into a section, have an **exact** de-crease twin, when going out of a (most natural -same) section. That was really hard to achieve, and why i made that tiny app i linked to above.
So the purpose was : Mirror-twins to get in and out exactly identically!
In that paradigm any change in the mirrored one, would be wrong.
So it actually come down to the initial heuristics.
Someone else could have completely new ideas, where a twin-copy could be used in some great creative way *"!if only those dumb bezier-handles could be used!"* 
But that situation is hypothetical, and imo the bezier implementation should not be halted because of that hypothetical usage.
That was the defense for my view *Handles does not need to be mirrored, only curves does*

**Comment by tresf** [comment]
@musikBear can you please explain what "can't be used" means?  I feel like we're trying to sugar coat a terrible decision here.

**Comment by codythecoder** [comment]
I don't know why I didn't think of this before but I think I have a solution. If all the control points are locked within the screen, except for the first on the left and the last on the right, the curve will always remain within the screen and the problem of flipping the x will also be removed. This will also keep the use case of this kind of curve:
![temp](https://cloud.githubusercontent.com/assets/13900631/10928112/ca04693c-82f9-11e5-8842-b3b42d9d23ae.png)

**Comment by musikBear** [comment]
> @musikBear can you please explain what "can't be used" means? I feel like we're trying to sugar coat a terrible decision here.

Do you refer to this sentence
> Someone else could have completely new ideas, where a twin-copy could be used in some great creative way "!if only those dumb bezier-handles could be used!"

eg "could be used!"
that is a point in that unknown perhaps non existing creative idea someone might think of, but **that** would need flipped handles, and then my heuristics would sabotage that creativity. But does it exists?
